pkg/pillar/cmd/baseosmgr/forcefallback.go

```diff
@@ -0,0 +1,170 @@
+}
+
+// handleForceFallback checks if we have a file with a ForceFallbackCounter
+// and if so, if the counter has changed.
+// If it has, then if the current partition is active, the other partition is
+// unused and the other partition has some non-empty EVE version string,
+// then it will mark the other partition as Updating.
+// This update to ZbootStatus will make nodeagent trigger a reboot.
+func handleForceFallback(ctxPtr *baseOsMgrContext, status types.ZedAgentStatus) {
+
+	counter, found := readForceFallbackCounter()
+	if !found {
+		// Just write the current value
+		log.Functionf("Saving initial ForceFallbackCounter %d",
+			status.ForceFallbackCounter)
+		writeForceFallbackCounter(status.ForceFallbackCounter)
+		return
+	}
+	if counter == status.ForceFallbackCounter {
+		log.Functionf("No change to ForceFallbackCounter %d", counter)
+		return
+	}
+	log.Noticef("Handle ForceFallbackCounter update from %d to %d",
+		counter, status.ForceFallbackCounter)
+
+	curPartName := zboot.GetCurrentPartition()
+	partStatus := getZbootStatus(ctxPtr, curPartName)
+	if partStatus == nil {
+		log.Warnf("No current partition status for %s; ignoring ForceFallback",
+			curPartName)
+		return
+	}
+	if partStatus.PartitionState != "active" {
+		log.Warnf("Current partition state %s not active; ignoring ForceFallback",
+			partStatus.PartitionState)
+		return
+	}
+	shortVerCurPart := partStatus.ShortVersion
+	otherPartName := zboot.GetOtherPartition()
+	partStatus = getZbootStatus(ctxPtr, otherPartName)
+	if partStatus == nil {
+		log.Warnf("No other partition status for %s; ignoring ForceFallback",
+			otherPartName)
+		return
+	}
+	shortVerOtherPart := partStatus.ShortVersion
+	if shortVerOtherPart == "" {
+		log.Warnf("Other partition has no version; ignoring ForceFallback")
+		return
+	}
+	if partStatus.PartitionState != "unused" {
+		log.Warnf("Other partition state %s not unused; ignoring ForceFallback",
+			partStatus.PartitionState)
+		return
+	}
+	log.Noticef("ForceFallback from %s to %s",
+		shortVerCurPart, shortVerOtherPart)
+
+	zboot.SetOtherPartitionStateUpdating(log)
+	updateAndPublishZbootStatus(ctxPtr,
+		partStatus.PartitionLabel, false)
+	baseOsStatus := lookupBaseOsStatusByPartLabel(ctxPtr, partStatus.PartitionLabel)
+	if baseOsStatus != nil {
+		baseOsSetPartitionInfoInStatus(ctxPtr, baseOsStatus,
+			partStatus.PartitionLabel)
+		publishBaseOsStatus(ctxPtr, baseOsStatus)
+	}
+	writeForceFallbackCounter(status.ForceFallbackCounter)
+}
+
+// readForceFallbackCounter reads the persistent file
+// If the file doesn't exist or doesn't contain an integer it returns false
+func readForceFallbackCounter() (int, bool) {
+	if _, err := os.Stat(forcefallbackFilename); err != nil {
+		return 0, false
+	}
+	s, err := read(forcefallbackFilename)
+	if err != nil {
+		return 0, false
+	}
+	c, err := strconv.Atoi(s)
+	if err != nil {
+		log.Errorf("readForceFallbackCouter parse %s failed: %s",
+			s, err)
+		return 0, false
```

Comment:
Does it make sense to write the file with a meaningful number at this point?

Reply:
Having a read function have the side effect of doing a write seems very odd to me. The fact that the file is missing or corrupt is returned to the caller as "false" and the caller can do what makes sense.

Reply:
Sure, I wasn't saying that the file should be written from inside this function. What I meant to ask is if this function should return something explicit that hints the callers about the non-integer value present in this file and if the callers should write a number, so that the next fallback request succeeds.

Reply:
The next fallback request will always write a number to the file.
Basically the approach is that a bad file (empty or non-empty) is semantically equivalent with no file when we read.
And when we write we always write the file.